README
NAME
    Video::Capture::V4l - Perl interface to the Video4linux
    framegrabber interface.

SYNOPSIS
      use Video::Capture::V4l;

DESCRIPTION
    Not documentation AGAIN! Please see the scripts grab, inexer or
    vbi that are packaged in the distribution and direct any
    question and feature requests (as well as bug reports) to the
    author.

Exported constants
    The following hideous constants are defined in the
    `Video::Capture::V4l' package, but you seldom need to use them.

      AUDIO_BASS
      AUDIO_MUTABLE
      AUDIO_MUTE
      AUDIO_TREBLE
      AUDIO_VOLUME
      CAPTURE_EVEN
      CAPTURE_ODD
      MAX_FRAME
      MODE_AUTO
      MODE_NTSC
      MODE_PAL
      MODE_SECAM
      PALETTE_COMPONENT
      PALETTE_GREY
      PALETTE_HI240
      PALETTE_PLANAR
      PALETTE_RAW
      PALETTE_RGB24
      PALETTE_RGB32
      PALETTE_RGB555
      PALETTE_RGB565
      PALETTE_UYVY
      PALETTE_YUV410P
      PALETTE_YUV411
      PALETTE_YUV411P
      PALETTE_YUV420
      PALETTE_YUV420P
      PALETTE_YUV422
      PALETTE_YUV422P
      PALETTE_YUYV
      SOUND_LANG1
      SOUND_LANG2
      SOUND_MONO
      SOUND_STEREO
      TUNER_LOW
      TUNER_MBS_ON
      TUNER_NORM
      TUNER_NTSC
      TUNER_PAL
      TUNER_RDS_ON
      TUNER_SECAM
      TUNER_STEREO_ON
      TYPE_CAMERA
      TYPE_TV
      VC_AUDIO
      VC_TUNER
      TYPE_CAPTURE
      TYPE_CHROMAKEY
      TYPE_CLIPPING
      TYPE_FRAMERAM
      TYPE_MONOCHROME
      TYPE_OVERLAY
      TYPE_SCALES
      TYPE_SUBCAPTURE
      TYPE_TELETEXT
      TYPE_TUNER

AUTHOR
    Marc Lehmann <pcg@goof.com>

LICENSE
    This module is available under GPL only (see the file COPYING
    for details), if you want an exception please contact the
    author, who might grant exceptions freely ;)

SEE ALSO
    perl(1).



Changes
Revision history for Perl extension Video::Capture::V4l.

0.901 Tue Jun  6 01:35:18 CEST 2006
	- fix lvalue-cast isn't lvalue bug in rtjpeg.c

0.9   Fri Nov  4 22:20:53 CET 2005
	- fix RTJpeg compilation flags.
        - expose decode_vps and decode_vt subfunctions.
        - fix CNI decoding.
        - updated VPS_CNI and 8/30 format 1 and format 2 codes,
          added X/26 codes, from TR-101-231 2005-10, and converted
          to utf-8.

0.225 Thu Mar  3 18:00:52 CET 2005
	- change of contact address.

0.224 Sun Jun 13 18:02:32 CEST 2004
	- fix some compile bugs, very very small documentation updates.
        - Video::Capture::V4l2 might replace this module in the future,
          details unclear.

0.222 Mon Feb 25 00:23:43 CET 2002
	- work around missing cancellation bug in linux by explicitly
          calling testcancel.

0.221 Thu May 11 18:28:29 CEST 2000
	- fixed oh so stupid bug reported by M C Lincoln.

0.22  Thu May 11 03:22:52 CEST 2000
	- changed VBI_VT format so that the full data is always returned in
          slot #3.
        - updated GPL to version 2.
        - minor fixes.

0.21  Sun Apr 16 11:13:04 CEST 2000
        - moved scripts to the example subdir, fixed a few bugs, writing an
          article in tpj ;)
        - added RTjpeg codec module.

0.20  Wed Dec  1 21:21:02 CET 1999
	- fixed segfault problems due to non-reentrant malloc on many systems
          (by pre-allocating lotsamemory of cousre ;).

0.11  Wed Nov 24 18:45:23 CET 1999
	- added epgview-simple (hack!), a regularly updated "coming next" screen.
        - added epgview, which requires Curses and is much nicer.
        - made it compile with perl-5.005_03.
        - removed epgview-simple again ;)
        - fixed "no dot in PATH" problem in V4l.

0.10  Sun Oct 24 21:18:00 CEST 1999
	- added Video::XawTV.
        - backlog(0) now cancels vbi frame grabbing.
        - added autotune.
        - the automatic re-centering was causing too many bit errors.

0.08  Sat Sep  4 19:17:42 CEST 1999
	- added dumpepg and renamed epg to getepg.
        - many bugfixes.
        - added VBI_OTHER (for other types of lines).
        - improved epg decoder & viewer.
        - fixed VPS/VT CNI code (damned %%&$& bit reversals).
        - vastly improved detection reliability.
        - added %VPS_CNI and %VT_NI tables.
        - preliminary premiere crypt data capturing.

0.04  Fri Aug 13 00:32:17 CEST 1999
	- added more example apps: vbi-info, epg.
        - vastly improved teletext decoding.

0.03  Wed Aug  4 21:15:28 CEST 1999
	- added VBI module and example videotext-decoder.
        - "cleaned" up file structure, as advised by Nick.
        - renamed Video::Capture::Frequencies to Video::Frequencies.

0.02  Wed Jul 14 14:21:37 CEST 1999
	- added the Video::Capture::Frequencies module.
        - added finder/indexer sample applications.
        - the frequency formula was bogus.

0.01  Thu Jul  1 20:32:32 1999
	- original version; created by h2xs 1.19



V4l/V4l.pm
package Video::Capture::V4l;

use strict 'subs';
use Carp;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK $AUTOLOAD);

require Exporter;
require DynaLoader;

use Fcntl;

$VERSION = '0.901';

@ISA = qw(Exporter DynaLoader);

@EXPORT = qw(
	MODE_AUTO MODE_NTSC MODE_PAL MODE_SECAM
	PALETTE_GREY PALETTE_HI240 PALETTE_PLANAR PALETTE_RAW PALETTE_RGB24
	PALETTE_RGB32 PALETTE_RGB555 PALETTE_RGB565 PALETTE_UYVY PALETTE_YUV410P
	PALETTE_YUV411 PALETTE_YUV411P PALETTE_YUV420 PALETTE_YUV420P PALETTE_YUV422
	PALETTE_YUV422P PALETTE_YUYV
	SOUND_LANG1 SOUND_LANG2 SOUND_MONO SOUND_STEREO
        VBI_MAXLINES VBI_BPL VBI_BPF
);

@EXPORT_OK = qw(
	MODE_AUTO MODE_NTSC MODE_PAL MODE_SECAM
	PALETTE_GREY PALETTE_HI240 PALETTE_PLANAR PALETTE_RAW PALETTE_RGB24
	PALETTE_RGB32 PALETTE_RGB555 PALETTE_RGB565 PALETTE_UYVY PALETTE_YUV410P
	PALETTE_YUV411 PALETTE_YUV411P PALETTE_YUV420 PALETTE_YUV420P PALETTE_YUV422
	PALETTE_YUV422P PALETTE_YUYV
	SOUND_LANG1 SOUND_LANG2 SOUND_MONO SOUND_STEREO
	AUDIO_MUTE 

	AUDIO_BASS AUDIO_MUTABLE AUDIO_TREBLE AUDIO_VOLUME
	TUNER_LOW TUNER_MBS_ON TUNER_NORM TUNER_NTSC TUNER_PAL
	TUNER_RDS_ON TUNER_SECAM TUNER_STEREO_ON
	VC_AUDIO VC_TUNER
	TYPE_CAMERA TYPE_TV

	BASE_VIDIOCPRIVATE
	CAPTURE_EVEN
	CAPTURE_ODD
	CLIPMAP_SIZE
	CLIP_BITMAP
	MAX_FRAME
	NO_UNIT
	PALETTE_COMPONENT
	WINDOW_INTERLACE
	HARDWARE_AZTECH
	HARDWARE_BROADWAY
	HARDWARE_BT848
	HARDWARE_CADET
	HARDWARE_GEMTEK
	HARDWARE_PERMEDIA2
	HARDWARE_PLANB
	HARDWARE_PMS
	HARDWARE_PSEUDO
	HARDWARE_QCAM_BW
	HARDWARE_QCAM_C
	HARDWARE_RIVA128
	HARDWARE_RTRACK
	HARDWARE_RTRACK2
	HARDWARE_SAA5249
	HARDWARE_SAA7146
	HARDWARE_SF16MI
	HARDWARE_TYPHOON
	HARDWARE_VIDEUM
	HARDWARE_VINO
	HARDWARE_ZOLTRIX
	TYPE_CAPTURE
	TYPE_CHROMAKEY
	TYPE_CLIPPING
	TYPE_FRAMERAM
	TYPE_MONOCHROME
	TYPE_OVERLAY
	TYPE_SCALES
	TYPE_SUBCAPTURE
	TYPE_TELETEXT
	TYPE_TUNER
);

# shit
sub VBI_MAXLINES	(){ 32 }
sub VBI_BPL		(){ 2048 }
sub VBI_BPF		(){ VBI_BPL * VBI_MAXLINES }

bootstrap Video::Capture::V4l $VERSION;

sub new(;$) {
   my $class  = shift;
   my $device = shift || "/dev/video0";
   my $self = bless { device => $device }, $class;

   $self->{handle} = local *{$device};
   sysopen $self->{handle},$device,O_RDWR or return;
   $self->{fd} = fileno ($self->{handle});
   $self->{capability} = _capabilities_new ($self->{fd});
   $self->{picture} = _picture_new ($self->{fd});

   $self->{capability}->get
   && $self->{picture}->get ? $self : ();
}

sub capability($) { shift->{capability} }
sub picture($)    { shift->{picture} }

sub channel($$) {
   my $c = _channel_new ($_[0]->{fd});
   $c->channel ($_[1]);
   $c->get ? $c : ();
}

sub tuner($$) {
   my $c = _tuner_new ($_[0]->{fd});
   $c->tuner ($_[1]);
   $c->get ? $c : ();
}

sub audio($$) {
   my $c = _audio_new ($_[0]->{fd});
   $c->audio ($_[1]);
   $c->get ? $c : ();
}

sub freq($;$) {
   _freq (shift->{fd},@_);
}

package Video::Capture::V4l::Capability;

sub capture   ($){ shift->type & &Video::Capture::V4l::TYPE_CAPTURE   }
sub tuner     ($){ shift->type & &Video::Capture::V4l::TYPE_TUNER     }
sub teletext  ($){ shift->type & &Video::Capture::V4l::TYPE_TELETEXT  }
sub overlay   ($){ shift->type & &Video::Capture::V4l::TYPE_OVERLAY   }
sub chromakey ($){ shift->type & &Video::Capture::V4l::TYPE_CHROMAKEY }
sub clipping  ($){ shift->type & &Video::Capture::V4l::TYPE_CLIPPING  }
sub frameram  ($){ shift->type & &Video::Capture::V4l::TYPE_FRAMERAM  }
sub scales    ($){ shift->type & &Video::Capture::V4l::TYPE_SCALES    }
sub monochrome($){ shift->type & &Video::Capture::V4l::TYPE_MONOCHROME}
sub subcapture($){ shift->type & &Video::Capture::V4l::TYPE_SUBCAPTURE}

package Video::Capture::V4l::Channel;

sub tuner     ($){ shift->flags & &Video::Capture::V4l::VC_TUNER      }
sub audio     ($){ shift->flags & &Video::Capture::V4l::VC_AUDIO      }

sub tv        ($){ shift->type & &Video::Capture::V4l::TYPE_TV        }
sub camera    ($){ shift->type & &Video::Capture::V4l::TYPE_CAMERA    }

package Video::Capture::V4l::Tuner;

sub pal       ($){ shift->flags & &Video::Capture::V4l::TUNER_PAL     }
sub ntsc      ($){ shift->flags & &Video::Capture::V4l::TUNER_NTSC    }
sub secam     ($){ shift->flags & &Video::Capture::V4l::TUNER_SECAM   }
sub low       ($){ shift->flags & &Video::Capture::V4l::TUNER_LOW     }
sub norm      ($){ shift->flags & &Video::Capture::V4l::TUNER_NORM    }
sub stereo_on ($){ shift->flags & &Video::Capture::V4l::TUNER_STEREO_ON}
sub rds_on    ($){ shift->flags & &Video::Capture::V4l::TUNER_RDS_ON  }
sub mbs_on    ($){ shift->flags & &Video::Capture::V4l::TUNER_MBS_ON  }

package Video::Capture::V4l::Audio;

sub mute      ($){ shift->flags & &Video::Capture::V4l::AUDIO_MUTE    }
sub mutatble  ($){ shift->flags & &Video::Capture::V4l::AUDIO_MUTABLE }
sub volume    ($){ shift->flags & &Video::Capture::V4l::AUDIO_VOLUME  }
sub bass      ($){ shift->flags & &Video::Capture::V4l::AUDIO_BASS    }
sub treble    ($){ shift->flags & &Video::Capture::V4l::AUDIO_TREBLE  }

package Video::Capture::V4l::VBI;

use Fcntl;

sub new(;$) {
   my $class  = shift;
   my $device = shift || "/dev/vbi0";
   my $self = bless { device => $device }, $class;

   $self->{handle} = local *{$device};
   sysopen $self->{handle},$device,O_RDWR or return;
   $self->{fd} = fileno ($self->{handle});

   $self
}

sub fileno($) {
   $_[0]->{fd}
}

1;
__END__

=head1 NAME

Video::Capture::V4l - Perl interface to the Video4linux framegrabber interface.

=head1 SYNOPSIS

  use Video::Capture::V4l;

=head1 DESCRIPTION

Not documentation AGAIN! Please see the scripts grab, inexer or vbi that
are packaged in the distribution and direct any question and feature
requests (as well as bug reports) to the author.

=head1 Exported constants

The following hideous constants are defined in the C<Video::Capture::V4l> package,
but you rarely need to use them.

  AUDIO_BASS
  AUDIO_MUTABLE
  AUDIO_MUTE
  AUDIO_TREBLE
  AUDIO_VOLUME
  CAPTURE_EVEN
  CAPTURE_ODD
  MAX_FRAME
  MODE_AUTO
  MODE_NTSC
  MODE_PAL
  MODE_SECAM
  PALETTE_COMPONENT
  PALETTE_GREY
  PALETTE_HI240
  PALETTE_PLANAR
  PALETTE_RAW
  PALETTE_RGB24
  PALETTE_RGB32
  PALETTE_RGB555
  PALETTE_RGB565
  PALETTE_UYVY
  PALETTE_YUV410P
  PALETTE_YUV411
  PALETTE_YUV411P
  PALETTE_YUV420
  PALETTE_YUV420P
  PALETTE_YUV422
  PALETTE_YUV422P
  PALETTE_YUYV
  SOUND_LANG1
  SOUND_LANG2
  SOUND_MONO
  SOUND_STEREO
  TUNER_LOW
  TUNER_MBS_ON
  TUNER_NORM
  TUNER_NTSC
  TUNER_PAL
  TUNER_RDS_ON
  TUNER_SECAM
  TUNER_STEREO_ON
  TYPE_CAMERA
  TYPE_TV
  VC_AUDIO
  VC_TUNER
  TYPE_CAPTURE
  TYPE_CHROMAKEY
  TYPE_CLIPPING
  TYPE_FRAMERAM
  TYPE_MONOCHROME
  TYPE_OVERLAY
  TYPE_SCALES
  TYPE_SUBCAPTURE
  TYPE_TELETEXT
  TYPE_TUNER

=head1 AUTHOR

Marc Lehmann <schmorp@schmorp.de>

=head1 LICENSE

This module is available under GPL only (see the file COPYING for
details), if you want an exception please contact the author, who might
grant exceptions freely ;)

=head1 SEE ALSO

perl(1).

=cut
